### DISCUSSION TITLE
i want create window but i dont want any Silk.NET.Windowing.Glfw or  Silk.NET.Windowing.SDL,How ?

### DISCUSSION THREAD

**Discussion by sgf** [dotnet/Silk.NET#1820] [Q&A]
Since I want to develop a D2d program, I shouldn't have any need to bring Glfw and SDL, right? 
Why force developers to choose one of the two here?
Isn't it good to directly encapsulate the native win32api?

**Comment by Perksey** [comment]
It is impossible to create a window without bringing in a windowing implementation, Windowing.Common is just abstractions. You could create your own Windowing.Common implementation using User32 if you wanted, but it’s far easier to just use one of our precanned implementations.

**Reply by sgf** [reply]
However, these two bring unclean factors to this library and add unnecessary native dll dependencies.

**Reply by Perksey** [reply]
Then it seems like you will need to do it all yourself using User32 given that you don’t want to use the precanned implementations.

**Reply by sgf** [reply]
It's not that I don't want to use the pre-implementation, but the pre-implementation relies on external third-party dlls.

**Comment by sgf** [comment]
Anyway, thanks for the answer.
In fact, I tried the encapsulation of this library. The encapsulation of this library is very low-level, and it almost just copies the C++ code. Moreover, due to the limitations of the C# language, the same code has fewer lines in C++.
Maybe I should abandon this library. I don't feel any advantages, maybe better performance is its advantage. But I don't think it is more convenient to use.

**Reply by Perksey** [reply]
If you think overall that your entire program when completed will have less lines if implemented in C++ instead of C# and is more efficient to develop in, then yes I encourage you to use C++.

**Reply by sgf** [reply]
No, you misunderstood again, I was talking about other libraries wrapping it better. Not using C++ directly is better.

**Reply by Perksey** [reply]
If having one more native binary to interoperate with or writing a few more lines of infrastructure code you’ll never touch again is enough to preclude us as an option then I encourage you to use an alternative library that is more suited to your desires. You don’t have to use Silk.NET, and I don’t gain anything by you using Silk.NET - quite the contrary, actually.

**Reply by sgf** [reply]
It's not that the infrastructure layer is bad, but that this library uses too many pointers directly. Although I also have experience in pointer writing, C# itself does not good support pointers(There are many restrictions).
For example: CS0212: You can only take the address of an unfixed expression inside of a fixed statement initializer.
This results in that when calling a function, a temporary variable must be used for transfer, and then the pointer is obtained and then assigned to the pointer variable within the class.

**Reply by Beyley** [reply]
We have many overloads which dont require using raw pointers, can you give specific examples you are having problems with?

**Reply by sgf** [reply]
```C#
            var guid = ID2D1Factory.Guid;
            ID2D1Factory* pD2D1Factory = (ID2D1Factory*)IntPtr.Zero;
            var hr = d2d.D2D1CreateFactory(FactoryType.SingleThreaded, ref guid, new FactoryOptions(DebugLevel.Error), ((void**)&pD2D1Factory));
            SilkMarshal.ThrowHResult(hr);
            _pD2D1Factory = pD2D1Factory;
```
 
At the same time, there seems to be something missing in the API, ColorF cannot be directly converted to D3Dcolorvalue, although I implemented it myself.
```

       //clear the render target contents
        m_pRT->Clear(D2D1::ColorF(D2D1::ColorF::Black));

```

**Reply by sgf** [reply]
```C#
    public record struct IColor(byte Red, byte Green, byte Blue, byte Alpha=255)
    {
        const int sc_alphaShift = 24;
        const int sc_redShift = 16;
        const int sc_greenShift = 8;
        const int sc_blueShift = 0;

        public IColor(Silk.NET.Direct2D.Enum @enum) : this((int)@enum) { }

        public IColor(int argb):this(
            (byte)((argb >> sc_redShift) & 0xff),
            (byte)((argb >> sc_greenShift) & 0xff),
            (byte)((argb >> sc_blueShift) & 0xff),
            (byte)((argb >> sc_alphaShift) & 0xff)) {
        }

        public static implicit operator D3Dcolorvalue(IColor color)
        {
            return new D3Dcolorvalue(
                  color.Red / 255f,
                  color.Green / 255f,
                  color.Blue / 255f,
                  color.Alpha / 255f);
        }

    }
```

**Reply by Beyley** [reply]
Fir the ComPtr part, you can (and probably should) use our `ComPtr<T>` wrapper, you can see how to use them [here](https://github.com/dotnet/Silk.NET/blob/main/examples/CSharp/Direct3D11%20Tutorials/Tutorial%201.2%20-%20Hello%20quad/Program.cs)

**Reply by sgf** [reply]
Well, I must thank you for pointing out my problem. At that time, I looked at the case without D2D, so I didn't look carefully.
